Server install: Nomad service discovery

paladin-devops opened the pull request:
This PR updates the server installation for Nomad to support service discovery for Nomad or Consul by specifying the service discovery provider as a flag for the `waypoint install` command. 

Closes #3376.

NOTE: The old config, `-nomad-consul-service`, is still supported with this PR update, but it should be phased out eventually. The new `-nomad-service-provider` config supports pre- and post-1.3.0 Waypoint installs to Nomad.

paladin-devops committed:
core/runner: Add runner install command to Waypoint

paladin-devops committed:
Add option to Nomad install for Waypoint server to support Nomad or Consul as a service provider.

paladin-devops committed:
Changelog & docs.

briancain commented on internal/serverinstall/nomad.go:
```diff
@@ -1394,11 +1428,11 @@ func (i *NomadInstaller) UpgradeFlags(set *flag.Set) {
+		Target:  &i.config.serviceProvider,
 		Usage:   "Create service for Waypoint UI and Server in Consul.",
-		Default: false, // default to false, make sure people opt into this for upgrades
+		Default: "consul",
```
This could be a good case for an `EnumVar` so we only ever allow `nomad` or `consul` and can fail early if they give us an unsupported option!

paladin-devops replied:
@briancain in order to still support an install method without a service discovery provider (pre-Nomad 1.3.0), I'll probably name this 3rd option (no service discovery provider) `none`.

briancain commented on internal/serverinstall/nomad.go:
```diff
@@ -285,17 +299,21 @@ func (i *NomadInstaller) Install(
+			"This could change in the future if Nomad creates a new allocation for the Waypoint server,\n" +
+			"which would break all existing Waypoint contexts.\n\n" +
+			"It is recommended to use Consul for determining Waypoint servers IP running on Nomad rather than\n" +
+			"relying on the static IP that is initially set up for this allocation.")
```
Hmm, these three outputs might be better suited for `ui.Output` instead of a step group since they are single output! I think @evanphx has mentioned to be before that doing `\n` new lines in step groups might be a bad idea too.

evanphx replied:
We were doing it previously, so I think it's fine. If it's weird we should probably just fix it so it works right too.

briancain reviewed (changes requested):
Haven't tried the full install path for these yet, but I have a few minor code comments! Changes make sense to me otherwise 👍🏻

evanphx commented on internal/serverinstall/nomad.go:
```diff
@@ -1394,11 +1428,11 @@ func (i *NomadInstaller) UpgradeFlags(set *flag.Set) {
-		Name:    "nomad-consul-service",
-		Target:  &i.config.consulService,
+	set.StringVar(&flag.StringVar{
+		Name:    "nomad-service-provider",
```
Might be a good idea to support the old flag for a release or two rather than remove it immediately.

paladin-devops replied:
@evanphx updated!

paladin-devops committed:
Use ui.Output instead of step groups for service discovery provider info on Nomad install.

paladin-devops committed:
Revert removal of -nomad-consul-service flag. Add support for both the old and new flags to configure service discovery.

paladin-devops committed:
Fix service provider option to be an EnumSingleVar.

paladin-devops committed:
Fix behavior of service discovery to support default values of fresh installs and upgrades.

demophoon commented on internal/serverinstall/nomad.go:
```diff
@@ -813,29 +839,48 @@ func waypointNomadJob(c nomadConfig, rawRunFlags []string) *api.Job {
 	// One service added for Waypoint UI, and one for Waypoint backend port
-	if c.consulService {
+	var services []*api.Service
+	if (c.serviceProvider == "consul" && (c.consulService || upgrade)) || (c.consulService && (c.serviceProvider == "consul" || upgrade)) {
```
Would it be worth while to emit a warning if we find that a user is using the `-nomad-consul-service` flag instead of the `-nomad-service-provider` flag?

demophoon replied:
Oh, i see, its above :D ignore me!

paladin-devops committed:
Add flags to Nomad install to support service provider-agnostic tags, and prefer Consul tag flags for backwards compatibility.

paladin-devops commented on internal/serverinstall/nomad.go:
```diff
@@ -319,11 +332,11 @@ func (i *NomadInstaller) Upgrade(
-		sw := sg.Add("Nomad Consul Service is disabled. If you had previously enabled " +
-			"it in the last installation, please stop this upgrade and re-run with -nomad-consul-service=true.")
+		sw := sg.Add("Service discovery is disabled for the Waypoint Nomad job. If you had previously enabled " +
+			"it in the last installation, please stop this upgrade and re-run with -nomad-service-provider flag.")
```
@demophoon after giving this some thought, I think an interactive terminal event could be useful here. If the user is OK with proceeding, we let them, otherwise they can abort!

```suggestion
		proceed, err := opts.UI.Input(&terminal.Input{
			Prompt: "Service discovery is disabled for the Waypoint Nomad job. If you had previously enabled " +
				"it in the last installation, please stop this upgrade and re-run with the -nomad-service-provider flag. " +
				"Otherwise, enter 'yes' to continue the upgrade: ",
			Style:  "",
			Secret: false,
		})
		if err != nil {
			opts.UI.Output(
				"Error uninstalling server: %s",
				clierrors.Humanize(err),
				terminal.WithErrorStyle(),
			)
		} else if strings.ToLower(proceed) != "yes" {
			return nil, err
		}
```

demophoon replied:
This is a good solution to this particular concern!

paladin-devops committed:
Add prompt for user input during server upgrade for Nomad.

This prompt is added so that a user doesn't inadvertently remove service discovery from their Waypoint installation.

catsby commented on internal/serverinstall/nomad.go:
```diff
@@ -813,29 +853,61 @@ func waypointNomadJob(c nomadConfig, rawRunFlags []string) *api.Job {
 	// One service added for Waypoint UI, and one for Waypoint backend port
-	if c.consulService {
+	var services []*api.Service
+	if (c.serviceProvider == "consul" && (c.consulService || upgrade)) || (c.consulService && (c.serviceProvider == "consul" || upgrade)) {
```
This line is confusing for me so I tried to simplify it. I haven't been successful yet but I did find an edge case I wanted to ask about; could `serviceProvider` be `"nomad"`, `consulService` be `true`, while also `upgrade` is `true`? If so, then this top block will execute and not the `if c.serviceProvider == "nomad"` one.

See https://go.dev/play/p/3lLyP9eqHa1

paladin-devops replied:
@catsby yes, the top block would indeed execute. However since `consulService` is false by default _for upgrades only_, a user would have to explicitly set `consulService` to true, & set `serviceProvider` to `nomad`. In light of @evanphx's earlier feedback for supporting the old configuration options, I felt that it was "okay" to favor the old flag being set over the new one.

Maybe it would make sense to print out some kind of warning or prompt for input if the input options are colliding...
```go
if c.consulService && c.serviceProvider == "nomad" {
  // warn the user this combo isn't valid
}
```

catsby replied:
I don't have operational experience here so I simply posed the question of whether it's valid or not after I noticed the possibility. I'll defer to you for the decision on if it's correct behavior or not 👍

catsby commented on internal/serverinstall/nomad.go:
```diff
@@ -813,29 +853,61 @@ func waypointNomadJob(c nomadConfig, rawRunFlags []string) *api.Job {
-	if c.consulService {
+	var services []*api.Service
+	if (c.serviceProvider == "consul" && (c.consulService || upgrade)) || (c.consulService && (c.serviceProvider == "consul" || upgrade)) {
 		token := ""
 		if c.consulToken == "" {
 			token = os.Getenv("CONSUL_HTTP_TOKEN")
 		} else {
 			token = c.consulToken
 		}
 		job.ConsulToken = &token
```
```suggestion
		token := c.consulToken
		if token == "" {
			token = os.Getenv("CONSUL_HTTP_TOKEN")
		}
		job.ConsulToken = &token

```

I'm not sure if we should check if token is empty before storing it in `job.ConsulToken` or if it's fine to put something empty in there ¯\\_(ツ)_/¯

catsby reviewed (commented):
Some nits and a question about some settings regarding service providers

paladin-devops committed:
Refactor retrieval of Waypoint address into a separate function.

paladin-devops commented on internal/serverinstall/nomad.go:
```diff
@@ -491,13 +518,26 @@ func (i *NomadInstaller) Upgrade(
 			addr.Addr = fmt.Sprintf("%s:%d", i.config.consulServiceHostname, grpcPort)
 			httpAddr = fmt.Sprintf("%s:%d", i.config.consulServiceHostname, httpPort)
 		}
+	} else if i.config.serviceProvider == "nomad" {
```
So refactored!

paladin-devops commented on internal/serverinstall/nomad.go:
```diff
@@ -319,13 +334,25 @@ func (i *NomadInstaller) Upgrade(
+			Style:  "",
+			Secret: false,
+		})
+		if err != nil {
+			opts.UI.Output(
+				"Error upgrading server: %s",
+				clierrors.Humanize(err),
+				terminal.WithErrorStyle(),
+			)
+		} else if strings.ToLower(proceed) != "yes" {
+			return nil, err
+		}
```
You summarized it correctly, but I think this needs a change! We should return the error if there was an error getting the input.

```suggestion
		if err != nil {
			opts.UI.Output(
				"Error upgrading server: %s",
				clierrors.Humanize(err),
				terminal.WithErrorStyle(),
			)
			return nil, err
		} else if strings.ToLower(proceed) != "yes" {
			return nil, err
		}
```

paladin-devops replied:
You summarized this correctly, but it should be changed!

```suggestion
		if err != nil {
			opts.UI.Output(
				"Error upgrading server: %s",
				clierrors.Humanize(err),
				terminal.WithErrorStyle(),
			)
			return nil, err
		} else if strings.ToLower(proceed) != "yes" {
			return nil, errors.New("upgrade aborted")
		}
```

paladin-devops committed:
Fix error handling for terminal input during Nomad upgrade.

briancain reviewed (approved):
🔍 Looks good to me code wise 👍🏻

xiaolin-ninja commented on internal/serverinstall/nomad.go:
```diff
@@ -286,19 +255,18 @@ func (i *NomadInstaller) Install(
 
-	if i.config.consulService {
-		s = sg.Add("The CLI has been configured to automatically install a Consul service for\n" +
+	if i.config.serviceProvider == "consul" && i.config.consulService {
```
I recall from a code review meeting that this was supposed to be an "or" instead of an "and"?

paladin-devops replied:
@xiaolin-ninja since the default values are "consul" and `true` for `serviceProvider` and `consulService`, respectively, we want this to be "and" because if the user sets something other than "consul" for `serviceProvider`, or sets `consulService` to false, we then do not want to create the job with a Consul service.

catsby reviewed (changes requested):
Please do not merge this until a patch for https://github.com/hashicorp/waypoint/issues/3601 is merged and released

catsby commented on internal/serverinstall/nomad.go:
```diff
@@ -237,42 +241,7 @@ func (i *NomadInstaller) Install(
-		httpAddr = hAddr
-		addr.Addr = serverAddr
-	}
+	httpAddr, addr.Addr, err = i.getWaypointAddress(client, allocID)
```
We are not checking this `err`, we should, or use `_` if it's something we can ignore

catsby commented on internal/serverinstall/nomad.go:
```diff
@@ -477,42 +504,7 @@ func (i *NomadInstaller) Upgrade(
-		httpAddr = hAddr
-		addr.Addr = serverAddr
-	}
+	httpAddr, addr.Addr, err = i.getWaypointAddress(client, allocID)
```
We are not checking this `err`, we should, or use `_` if it's something we can ignore

paladin-devops committed:
Return errors if getting the Waypoint server address fails during Nomad install.

briancain commented:
You might need to rebase off `main` to get the integration tests to pass, btw!

paladin-devops commented:
@briancain so rebased! 😄
